serde-generate/src/ocaml.rs

```diff
@@ -0,0 +1,211 @@
+use crate::{
+    indent::{IndentConfig, IndentedWriter},
+    CodeGeneratorConfig,
+};
+use heck::SnakeCase;
+use include_dir::include_dir as include_directory;
+use serde_reflection::{ContainerFormat, Format, Named, Registry, VariantFormat};
+use std::{
+    collections::{BTreeMap},
+    io::{Result, Write},
+    path::PathBuf,
+};
+
+fn capitalize(s: &str) -> String {
+    let mut c = s.chars();
+    match c.next() {
+        None => String::new(),
+        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
+    }
+}
+
+pub struct CodeGenerator<'a> {
+    config: &'a CodeGeneratorConfig,
+    libraries: Vec<String>,
+}
+
+struct OCamlEmitter<'a, T> {
+    out: IndentedWriter<T>,
+    generator: &'a CodeGenerator<'a>,
+}
+
+impl<'a> CodeGenerator<'a> {
+
+    pub fn new(config: &'a CodeGeneratorConfig) -> Self {
+        if config.c_style_enums {
+            panic!("OCaml does not support generating c-style enums");
+        }
+        Self {
+            config,
+            libraries: config.external_definitions.keys().map(|k| k.to_string()).collect::<Vec<_>>(),
+        }
+    }
+
+    pub fn output(&self, out: &mut dyn Write, registry: &Registry) -> Result<()> {
+        let mut emitter = OCamlEmitter {
+            out: IndentedWriter::new(out, IndentConfig::Space(2)),
+            generator: self,
+        };
+        emitter.output_preamble()?;
+        for (name, format) in registry {
+            emitter.output_container(name, format)?;
+        }
+        Ok(())
+    }
+}
+
+impl<'a, T> OCamlEmitter<'a, T>
+where T: Write, {
+
+    fn output_preamble(&mut self) -> Result<()> {
+        for namespace in self.generator.libraries.iter() {
+            writeln!(self.out, "open {}\n", capitalize(namespace))?;
+        }
+        Ok(())
+    }
+
+    fn format(&self, format: &Format) -> String {
+        use Format::*;
+        match format {
+            Variable(_) => panic!("incorrect value"),
+            TypeName(s) => s.to_string(),
+            Unit => "unit".into(),
+            Bool => "bool".into(),
+            I8 => "Stdint.int8".into(),
+            I16 => "Stdint.int16".into(),
+            I32 => "int32".into(),
+            I64 => "int64".into(),
+            I128 => "Stdint.int128".into(),
+            U8 => "Stdint.uint8".into(),
+            U16 => "Stdint.uint16".into(),
+            U32 => "Stdint.uint32".into(),
+            U64 => "Stdint.uint64".into(),
+            U128 => "Stdint.uint128".into(),
+            F32 => panic!("float32 not implemented in ocaml"),
```

Comment:
Can't we use `float` also for single precision floats in OCaml (with a special serialization function)?

Reply:
yes I fixed that today, I can use Int32.bits_of_float